relayer/channel.go

```diff
@@ -22,881 +20,133 @@ func (c *Chain) CreateOpenChannels(
-			if err != nil {
-				return nil, false, err
-			}
-
-			out.Src = append(out.Src,
-				updateMsg,
-				chanCloseConfirm,
-			)
-			isLast = true
-		}
-	}
-	return out, isLast, nil
-}
-
-// FindMatchingChannel will determine if there already exists a channel between source and counterparty
-// that matches the channel identifiers being passed in as arguments.
-func FindMatchingChannel(ctx context.Context, source *Chain, srcPortID, dstPortID, order, version string) (string, bool) {
-	// TODO: add appropriate offset and limits, along with retries
-	channelsResp, err := source.ChainProvider.QueryChannels(ctx)
-	if err != nil {
-		source.log.Info(
-			"Querying channels failed",
-			zap.String("src_chain_id", source.ChainID()),
-			zap.Error(err),
-		)
-		return "", false
-	}
-
-	for _, channel := range channelsResp {
-		if IsMatchingChannel(source, channel, srcPortID, dstPortID, order, version) {
-			// unused channel found
-			return channel.ChannelId, true
-		}
-	}
-
-	return "", false
-}
-
-// IsMatchingChannel determines if given channel matches required conditions.
-func IsMatchingChannel(source *Chain, channel *chantypes.IdentifiedChannel, srcPortID, dstPortID, order, version string) bool {
-	return channel.Ordering == OrderFromString(order) &&
-		IsConnectionFound(channel.ConnectionHops, source.ConnectionID()) &&
-		channel.Version == version &&
-		channel.PortId == srcPortID &&
-		channel.Counterparty.PortId == dstPortID &&
-		(((channel.State == chantypes.INIT || channel.State == chantypes.TRYOPEN) && channel.Counterparty.ChannelId == "") ||
-			(channel.State == chantypes.OPEN && (channel.Counterparty.ChannelId == "" || channel.Counterparty.PortId == dstPortID)))
-}
-
-// IsConnectionFound determines if given connectionId is present in channel connectionHops list.
-func IsConnectionFound(connectionHops []string, connectionID string) bool {
-	for _, id := range connectionHops {
-		if id == connectionID {
-			return true
-		}
-	}
-	return false
+func (c *Chain) CloseChannel(ctx context.Context, dst *Chain, maxRetries uint64, timeout time.Duration, srcChanID, srcPortID string) error {
+	srcPathChain := pathChain{
+		provider: c.ChainProvider,
+		pathEnd:  processor.NewPathEnd(c.PathEnd.ChainID, c.PathEnd.ClientID, "", []processor.ChannelKey{}),
+	}
+	dstPathChain := pathChain{
+		provider: dst.ChainProvider,
+		pathEnd:  processor.NewPathEnd(dst.PathEnd.ChainID, dst.PathEnd.ClientID, "", []processor.ChannelKey{}),
+	}
+
+	// Timeout is per message. Two close channel handshake messages, allowing maxRetries for each.
+	processorTimeout := timeout * 2 * time.Duration(maxRetries)
+
+	processorCtx, processorCtxCancel := context.WithTimeout(ctx, processorTimeout)
+	defer processorCtxCancel()
+
+	return processor.NewEventProcessor().
+		WithChainProcessors(
+			srcPathChain.chainProcessor(c.log),
+			dstPathChain.chainProcessor(c.log)).
```

Comment:
The `).` can go to its own line for visual symmetry with the rest of the builder pattern here.